--- src/grammar/ursa.ohm-bundle.d.ts	2023-12-31 13:44:52.122820018 +0000
+++ src/grammar/ursa.ohm-bundle.d.ts.bak	2023-12-31 13:44:51.666817094 +0000
@@ -4,14 +4,33 @@
 import {
   BaseActionDict,
   Grammar,
-  IterationNode,
-  Node,
+  Node as NodeBase,
   NonterminalNode,
+  NonterminalNode as NonterminalNodeBase,
   Semantics,
   TerminalNode
 } from 'ohm-js';
 
-export interface UrsaActionDict<T> extends BaseActionDict<T> {
+interface NodeI<Operations> extends NodeBase {
+  child(idx: number): Node<Operations>;
+  children: Node<Operations>[];
+  asIteration(): IterationNode<Operations>;
+}
+export type Node<Operations> = NodeI<Operations> & Operations;
+
+export type IterationNode<Operations> = Node<Operations>;
+
+interface NonterminalNodeI<Args, Operations> extends NodeI<Operations> {
+  // A hack to enable type-checking of `.args`. This member is really only
+  // present in `this` of semantics action routines.
+  args: {a: Args};
+}
+export type NonterminalNode<Args, Operations> = NonterminalNodeI<Args, Operations> & Operations;
+
+export interface UrsaActionDict<T, Node, NonterminalNode, IterationNode> extends BaseActionDict<T> {
+  _terminal?: (this: NonterminalNode) => T;
+  _nonterminal?: (this: NonterminalNode, ...children: NonterminalNode[]) => T;
+  _iter?: (this: NonterminalNode, ...children: NonterminalNode[]) => T;
   Sequence?: (this: NonterminalNode, arg0: NonterminalNode, arg1: NonterminalNode) => T;
   sc?: (this: NonterminalNode, arg0: IterationNode | NonterminalNode, arg1: NonterminalNode | TerminalNode) => T;
   maybeCommaParen?: (this: NonterminalNode, arg0: IterationNode, arg1: TerminalNode) => T;
@@ -136,18 +155,19 @@
   unicodeSpaceSeparator?: (this: NonterminalNode, arg0: TerminalNode) => T;
 }
 
-export interface UrsaSemantics extends Semantics {
-  addOperation<T>(name: string, actionDict: UrsaActionDict<T>): this;
-  extendOperation<T>(name: string, actionDict: UrsaActionDict<T>): this;
-  addAttribute<T>(name: string, actionDict: UrsaActionDict<T>): this;
-  extendAttribute<T>(name: string, actionDict: UrsaActionDict<T>): this;
+interface UrsaSemanticsI<Node, NonterminalNode, IterationNode, Operations> extends Semantics {
+  (match: MatchResult): Operations;
+  addOperation<T>(name: string, actionDict: UrsaActionDict<T, Node, NonterminalNode, IterationNode>): this;
+  extendOperation<T>(name: string, actionDict: UrsaActionDict<T, Node, NonterminalNode, IterationNode>): this;
+  addAttribute<T>(name: string, actionDict: UrsaActionDict<T, Node, NonterminalNode, IterationNode>): this;
+  extendAttribute<T>(name: string, actionDict: UrsaActionDict<T, Node, NonterminalNode, IterationNode>): this;
 }
+export type UrsaSemantics<Node, NonterminalNode, IterationNode, Operations> = UrsaSemanticsI<Node, NonterminalNode, IterationNode, Operations> & Operations;
 
 export interface UrsaGrammar extends Grammar {
-  createSemantics(): UrsaSemantics;
-  extendSemantics(superSemantics: UrsaSemantics): UrsaSemantics;
+  createSemantics<Node, NonterminalNode, IterationNode, Operations>(): UrsaSemantics<Node, NonterminalNode, IterationNode, Operations>;
+  extendSemantics<Node, NonterminalNode, IterationNode, Operations>(superSemantics: UrsaSemantics<Node, NonterminalNode, IterationNode, Operations>): UrsaSemantics<Node, NonterminalNode, IterationNode, Operations>;
 }
 
 declare const grammar: UrsaGrammar;
 export default grammar;
-
