--- src/grammar/ursa.ohm-bundle.d.ts	2023-12-28 17:41:52.729303563 +0000
+++ src/grammar/ursa.ohm-bundle.d.ts.bak	2023-12-28 17:41:52.301302184 +0000
@@ -4,14 +4,57 @@
 import {
   BaseActionDict,
   Grammar,
-  IterationNode,
-  Node,
-  NonterminalNode,
+  IterationNode as IterationNodeBase,
+  Node as NodeBase,
+  NonterminalNode as NonterminalNodeBase,
   Semantics,
   TerminalNode
 } from 'ohm-js';
 
+import {ArkExp} from '../ark/interpreter.js';
+import {Environment, FreeVars, CompiledArk} from '../ark/compiler.js';
+import {Definition, KeyValue, Arguments} from '../ursa/compiler.js';
+
+export type UrsaActionArgs = {
+  env: Environment;
+  inLoop?: boolean;
+  inFn?: boolean;
+};
+
+export type UrsaSemanticsArgs = {
+  a: UrsaActionArgs;
+  indent: number;
+  indentSize: number;
+};
+
+export type UrsaOperations = {
+  toExp(a: UrsaActionArgs): ArkExp;
+  toLval(a: UrsaActionArgs): ArkExp;
+  toDefinition(a: UrsaActionArgs): Definition;
+  toKeyValue(a: UrsaActionArgs): KeyValue;
+  toArguments(a: UrsaActionArgs): Arguments;
+  boundVars: string[];
+  freeVars(a: UrsaActionArgs): FreeVars;
+  fmt(indent: number, indentSize: number): string;
+  symref(a: UrsaActionArgs): CompiledArk;
+};
+
+export type Node = NodeBase & UrsaOperations;
+interface UrsaIterationNode extends IterationNodeBase {
+  children: Node[];
+}
+export type IterationNode = UrsaIterationNode & UrsaOperations;
+
+interface UrsaNonterminalNode extends NonterminalNodeBase {
+  args: UrsaSemanticsArgs;
+  asIteration(): IterationNode;
+}
+export type NonterminalNode = UrsaNonterminalNode & UrsaOperations;
+
 export interface UrsaActionDict<T> extends BaseActionDict<T> {
+  _terminal?: (this: NonterminalNode) => T;
+  _nonterminal?: (this: NonterminalNode, ...children: NonterminalNode[]) => T;
+  _iter?: (this: NonterminalNode, ...children: NonterminalNode[]) => T;
   Sequence?: (this: NonterminalNode, arg0: NonterminalNode, arg1: NonterminalNode) => T;
   sc?: (this: NonterminalNode, arg0: IterationNode | NonterminalNode, arg1: NonterminalNode | TerminalNode) => T;
   maybeCommaParen?: (this: NonterminalNode, arg0: IterationNode, arg1: TerminalNode) => T;
@@ -136,12 +179,13 @@
   unicodeSpaceSeparator?: (this: NonterminalNode, arg0: TerminalNode) => T;
 }
 
-export interface UrsaSemantics extends Semantics {
+interface UrsaSemanticsBase extends Semantics {
   addOperation<T>(name: string, actionDict: UrsaActionDict<T>): this;
   extendOperation<T>(name: string, actionDict: UrsaActionDict<T>): this;
   addAttribute<T>(name: string, actionDict: UrsaActionDict<T>): this;
   extendAttribute<T>(name: string, actionDict: UrsaActionDict<T>): this;
 }
+export type UrsaSemantics = UrsaSemanticsBase & UrsaOperations;
 
 export interface UrsaGrammar extends Grammar {
   createSemantics(): UrsaSemantics;
@@ -150,4 +194,3 @@
 
 declare const grammar: UrsaGrammar;
 export default grammar;
-
